### PR TITLE
SCP-3566 Conformance test suite skeleton

### PR DESCRIPTION
**Pull request by thealmarty**
This is the first of a series of PRs that will set up the Plutus conformance test suite. The conformance test suite aims to provide an official and comprehensive test suite that checks that the behaviour of Plutus conforms with the specified behaviour. This PR contains:
- a skeleton set up for UPLC and TPLC evaluation tests, TPLC typechecking tests. No tests are running yet. They will only run after a generic test function is set up in another PR.
- the first batch of UPLC evaluation tests taken from RV (https://github.com/runtimeverification/plutus-core-semantics/tree/master/tests).

In the next few PRs:
- Set up a generic test function that users can input an implementation to test against for all tests.
- Set up tests for UPLC evaluation tests.
- More tests will be added. With more output formats possibly.
- More README content.
- Set the tests up to run against the Agda implementation.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by thealmarty** [comment]
Because of the large number of files, this may be easier to review if you checkout the branch and look at the `plutus-conformance` directory.

File: plutus-conformance/plutus-conformance.cabal [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  NOTICE
+maintainer: marty.stumpf@iohk.io
+author: Plutus Core Team
+synopsis: Conformance Test Suite for Plutus Core
+description:
+    Comprehensive Conformance Test Suite for Plutus Core.
+category: Language, Plutus, Conformance
+build-type: Simple
+extra-doc-files: README.md
+extra-source-files:
```
Comment (michaelpj) [inline plutus-conformance/plutus-conformance.cabal:16]: wrong, but should maybe include the test files? I think we can worry about that later

File: plutus-conformance/plutus-conformance.cabal [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                 -Wmissing-deriving-strategies
+
+library
+    import: lang
+    exposed-modules:
+    hs-source-dirs:
+        uplc/
+    other-modules:
+    build-depends:
+        aeson -any,
```
Comment (michaelpj) [inline plutus-conformance/plutus-conformance.cabal:46]: prune this list

File: plutus-core/README.md [hunk 3]
```diff
@@ -36,7 +36,7 @@ The Haskell package `plutus-core` implements a range of functionality for manipu
 
 ### Parser and pretty printer
 
-The lexer & parser are based on Alex & Happy and the pretty printer uses the `prettyprinter` package. Names (identifiers) are interned using uniques as per `PlutusCore.Name`. They are also parameterised with an attribute used differently in different stages.
```
Comment (michaelpj) [inline plutus-core/README.md:39]: there are a bunch of unrelated changes here at the end, not the end of the world, but tidier to do those separately.

File: plutus-conformance/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
+# Plutus Core Conformance Test Suite
+
+## Specification
```
Comment (michaelpj) [inline plutus-conformance/README.md:3]: wrong content

Comment (michaelpj) [reply plutus-conformance/README.md:3]: Let's at least have a basic readme, and probably we should attribute the test sources immediately.

File: plutus-conformance/plutus-conformance.cabal [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        th-lift-instances -any,
+        th-utilities -any,
+        time -any,
+        transformers -any,
+        unordered-containers -any,
+        witherable -any,
+        word-array -any,
+        cardano-crypto-class -any,
+
+test-suite uplc-eval-test
```
Comment (michaelpj) [inline plutus-conformance/plutus-conformance.cabal:103]: Given that we're going to want to instantiate the test suites multiple times, I think probably what we want is to use the *library* component to store all the test functions and then have very minimal test suites that just call the library functions with appropriate arguments, e.g.

```
-- Lib.hs
uplcEvaluationTests :: EvaluationFunction -> TestTree

-- Spec.hs
import Lib

main = defaultMain (uplcEvaluationTests haskellCekEvaluationFunction)
```

**Comment by michaelpj** [comment]
Oh yeah, did you want to do the organization of the test cases into subdirectories in this PR? I think it's probably a good time to do it?

File: plutus-core/README.md (hunk 3 above)
Comment (thealmarty) [reply plutus-core/README.md:39]: Moved them to https://github.com/input-output-hk/plutus/pull/4553.

**Comment by michaelpj** [comment]
I think (3) is fine. We could also have an expected file which prints the parse error!

**Comment by michaelpj** [comment]
I think the things in `constant` should just be in `builtin`: they're mostly testing builtins. I think `builtin` can include builtin types too, in which case it's all consistent.

Some of the tests which don't have `.expected` files look like they should parse...

File: plutus-conformance/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+- Costing conformance 
+- Coverage test
+
+<!-- ## Testing alternative implementation
+
+We provide a function  -->
+
+## Untyped Plutus Core Evaluation
+
+The `uplc-eval-test` test suite ensures that the input untyped plutus core programs evaluate to the expected output. The expected output is obtained by evaluating the programs using the CEK machine.
```
Comment (michaelpj) [inline plutus-conformance/README.md:24]: I think the point of this suite is to test against the expected _semantics_ of PLC. So the fact that we use the CEK machine in our implementation of that is irrelevant!

File: plutus-conformance/README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+### The CEK machine
```
Comment (michaelpj) [inline plutus-conformance/README.md:26]: Similarly, I don't think this is relevant here.

File: plutus-conformance/README.md (hunk 6 above)
Comment (michaelpj) [reply plutus-conformance/README.md:24]: Well, but we *won't* be testing them against the CEK machine, we'll be testing them against the *checked-in input*, which for all they know we could have derived by hand!

This will be different if/when we start providing tests where we generate inputs and evaluate them with our implementation and theirs. Then we really will be testing against the CEK machine.

Anyway, you can leave it in for now if you like!

**Comment by thealmarty** [comment]
> Some of the tests which don't have `.expected` files look like they should parse...

You're right. I'm gonna take your suggestion and add `.expected` files to everything that doesn't have one.

File: plutus-conformance/README.md (hunk 6 above)
Comment (thealmarty) [reply plutus-conformance/README.md:24]: I guess what I was trying to say was that the checked-in input was gotten from the CEK machine **at some point**. I don't want to be misleading so I'll comment that out for now.

**Comment by thealmarty** [comment]
I would like to merge this and add a separate golden test suite for adding missing output files. It would be tidier that way.

**Comment by thealmarty** [comment]
[JIRA ticket](https://input-output.atlassian.net/browse/SCP-3886) added for the golden test suite.
